contribs/emissions/src/main/java/org/matsim/contrib/emissions/WarmEmissionHandler.java

```diff
@@ -123,106 +135,110 @@ public void handleEvent(VehicleEntersTrafficEvent event) {
 		linkLeaveCnt++;
 
-		if(!this.linkenter.containsKey(vehicleId)){
+		if (!this.linkEnterMap.containsKey(vehicleId)) { // vehicle has NOT entered this link
 			int maxLinkLeaveFirstActWarnCnt = 3;
-			if(linkLeaveFirstActWarnCnt < maxLinkLeaveFirstActWarnCnt){
+			if (linkLeaveFirstActWarnCnt < maxLinkLeaveFirstActWarnCnt) {
 				logger.info("Vehicle " + vehicleId + " is ending its first activity of the day and leaving link " + linkId + " without having entered.");
 				logger.info("This is because of the MATSim logic that there is no link enter event for the link of the first activity");
 				logger.info("Thus, no emissions are calculated for this link leave event.");
 				if (linkLeaveFirstActWarnCnt == maxLinkLeaveFirstActWarnCnt) logger.warn(Gbl.FUTURE_SUPPRESSED);
 			}
 			linkLeaveFirstActWarnCnt++;
-		} else if (!this.linkenter.get(vehicleId).getFirst().equals(linkId)){
+		} else if (!this.linkEnterMap.get(vehicleId).getFirst().equals(linkId)) { // vehicle HAS entered a link but not THIS one
 			int maxLinkLeaveSomeActWarnCnt = 3;
-			if(linkLeaveSomeActWarnCnt < maxLinkLeaveSomeActWarnCnt){
+			if (linkLeaveSomeActWarnCnt < maxLinkLeaveSomeActWarnCnt) {
 				logger.warn("Vehicle " + vehicleId + " is ending an activity other than the first and leaving link " + linkId + " without having entered.");
 				logger.warn("This indicates that there is some inconsistency in vehicle use; please check your inital plans file for consistency.");
 				logger.warn("Thus, no emissions are calculated neither for this link leave event nor for the last link that was entered.");
 				if (linkLeaveSomeActWarnCnt == maxLinkLeaveSomeActWarnCnt) logger.warn(Gbl.FUTURE_SUPPRESSED);
 			}
 			linkLeaveSomeActWarnCnt++;
 		} else {
-			double enterTime = this.linkenter.get(vehicleId).getSecond();
-			double travelTime;
-			if(!this.vehicleLeavesTraffic.containsKey(vehicleId) || !this.vehicleEntersTraffic.containsKey(vehicleId)){
-				travelTime = leaveTime - enterTime;
-			}
-			else if(!this.vehicleLeavesTraffic.get(vehicleId).getFirst().equals(event.getLinkId())
-					    || !this.vehicleEntersTraffic.get(vehicleId).getFirst().equals(event.getLinkId())){
-
-				travelTime = leaveTime - enterTime;
-			} else {
-				double arrivalTime = this.vehicleLeavesTraffic.get(vehicleId).getSecond();
-				double departureTime = this.vehicleEntersTraffic.get(vehicleId).getSecond();
-				travelTime = leaveTime - enterTime - departureTime + arrivalTime;
-
-				// now that we subtracted the time spent on the link outside of the vehicle
-				// we delete these events (so that when the link gets crossed once more
-				// the departure/arrival time are not wrongly taken into account)
-				// Thanks @markusstraub for providing this
-				this.vehicleLeavesTraffic.remove(vehicleId);
-				this.vehicleEntersTraffic.remove(vehicleId);
-			}
+			// the vehicle traversed the entire link, DO calculate emissions IF this was not after an activity
```

Comment:
The following logic is also too complicated. It uses `this.linkEnterMap.get(vehicleId).getSecond()` and then tests `this.linkEnterMap.containsKey(vehicleId)` on the next line. 

You could remove the two cases for logging basic Matsim stuff above. Then down here, something like the following should happen (not tested)

```
if (this.linkEnterMap.contains(vehicleId)) {
    var tuple = this.linkEnterMap.get(vehicleId);
    if (tuple.getFirst().equals(linkId)) {
        Vehicle vehicle = VehicleUtils.findVehicle(vehicleId, scenario);
	if (vehicle != null) {
            emissionsCalculation(vehicleId, vehicle, link, leaveTime, travelTime);
	    this.linkEnterMap.remove(vehicleId);
        } else {
            handleNullVehicle(vehicleId);
        }
    }
}

````